### DISCUSSION TITLE
Using events to initialize and cleanup "Generic" objects' memoized attributes

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#8469] [Usage Questions]
I can't seem to find the right mix of events to handle this use-case.

We have a collection of ORM objects that all share some base classes:

    DeclaredTable = declarative_base()

    class _UtilityObject(object):
        # shared application mixins
        pass
    
    class _ApplicationObject(object):
        # application specific mixins
        pass

    class ExampleObject(DeclaredTable, _UtilityObject, _ApplicationObject):
        pass

There are a handful of objects, like `ExampleObject` that utilize memoization for expensive operations.  On first access, the value is cached into the instance; subsequent access utilizes the attribute.

I found a bug where some Update/Edit routines did not properly remove the memoized data on some objects.

I would like to standardize this practice into something that can better protect against these scenarios, and be better for testing.

I was hoping to use mapper events to fix this, such as:

    class _ApplicationObject(object):
        _memoized_attribs = None

    class Foo(DeclaredTable, _UtilityObject, _ApplicationObject):
        _memoized_attribs = (MemoizedAttribute("attrib_name", defaultvalue), ...)

This is inspired by the `__table_args__` functionality. 

Ideally I would then do the following:

1. `after_configured(mapper)`, iterate all the mapped classes and set each memoized attribute to the default.

2. `after_flush(object)`, reset the memoized attributes for that class

Problems: The `sa.orm.mapper` doesn't seem to have access to any mapped classes.  The `after_flush` only seems to listen to Session (which makes sense based on the docs, but not the example)

Can anyone suggest the proper events, or how to properly access the data I need during these events?

**Comment by zzzeek** [comment]
after_configured() should work, can you explain "doesn't seem to have access to any mapped classes" ?

for after_flush(), you need to iterate through the sessions' .new and .dirty collections to locate objects that would have been in that flush.

**Reply by jvanasco** [reply]
> after_configured() should work, can you explain "doesn't seem to have access to any mapped classes" ?

I can't find anything in sqlalchemy.orm.Mapper. I also tried using sqlalchemy.orm.mapper, but that seems to only create a new mapper that has nothing in it.

I've got a (possibly?) temporary fix going by just examining all the objects in the namespace and running an initialization routine.  Knowing how to properly use the mapper would be great, but is not an immediate need.

> for after_flush(), you need to iterate through the sessions' .new and .dirty collections to locate objects that would have been in that flush.

Thanks! This part works perfectly now.

**Reply by zzzeek** [reply]
> > after_configured() should work, can you explain "doesn't seem to have access to any mapped classes" ?
> 
> I can't find anything in sqlalchemy.orm.Mapper. I also tried using sqlalchemy.orm.mapper, but that seems to only create a new mapper that has nothing in it.

oh it's for all mappers and not for a specific mapper you mean?   mapper_configured() might be better

https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=after_configured#sqlalchemy.orm.MapperEvents.mapper_configured

> 
> I've got a (possibly?) temporary fix going by just examining all the objects in the namespace and running an initialization routine. Knowing how to properly use the mapper would be great, but is not an immediate need.
> 
> > for after_flush(), you need to iterate through the sessions' .new and .dirty collections to locate objects that would have been in that flush.
> 
> Thanks! This part works perfectly now.

ok
